modules/javafx.graphics/src/main/java/javafx/scene/CssStyleHelper.java

```diff
@@ -564,20 +565,25 @@ private Set<PseudoClass>[] getTransitionStates(final Node node) {
 
         int count = 0;
         parent = node;
+
         while (parent != null) { // This loop traverses through all ancestors till root
-            final CssStyleHelper helper = (parent instanceof Node) ? parent.styleHelper : null;
-            if (helper != null) {
-                final Set<PseudoClass> pseudoClassState = parent.pseudoClassStates;
-                retainedStates[count] = new PseudoClassState();
-                retainedStates[count].addAll(pseudoClassState);
-                // retainAll method takes the intersection of pseudoClassState and helper.triggerStates
-                retainedStates[count].retainAll(helper.triggerStates);
-                count += 1;
+            if (parent.styleHelper != null) {
+                PseudoClassState pseudoClassState = new PseudoClassState();
+
+                pseudoClassState.addAll(parent.pseudoClassStates);
+                pseudoClassState.retainAll(parent.styleHelper.triggerStates);
+
+                retainedStates[count++] = ImmutablePseudoClassSetsCache.of(pseudoClassState);
             }
+
             parent = parent.getParent();
         }
 
-        final Set<PseudoClass>[] transitionStates = new PseudoClassState[count];
+        if (count == depth) {
```

Comment:
Can you explain why we can return here?

Reply:
Yes, we want to return a `Set[]` which has pseudo class state information for each nesting level of `Node`, as each entry in the array represents a nesting level (0 = `Node`, 1 = parent, 2 = parent.parent, etc.).  We allocate the array with the maximum possible nesting level (`depth`).  The loop then runs to fill in each level.  It's possible however that a level is unused, and in that case the array would be too long, so it must be trimmed.   If however all levels were used, there's no need to reallocate the array and we can return early.